jsettlers.logic/src/main/java/jsettlers/logic/movable/BehaviorTreeFactory.java

```diff
@@ -0,0 +1,116 @@
+import jsettlers.logic.movable.simplebehaviortree.NodeStatus;
+import jsettlers.logic.movable.simplebehaviortree.Tick;
+import jsettlers.logic.movable.simplebehaviortree.nodes.Action;
+import jsettlers.logic.movable.simplebehaviortree.nodes.Condition;
+import jsettlers.logic.movable.simplebehaviortree.nodes.Guard;
+import jsettlers.logic.movable.simplebehaviortree.nodes.Inverter;
+import jsettlers.logic.movable.simplebehaviortree.nodes.Selector;
+import jsettlers.logic.movable.simplebehaviortree.nodes.Sequence;
+import jsettlers.logic.movable.simplebehaviortree.nodes.Succeeder;
+import jsettlers.logic.movable.simplebehaviortree.nodes.WaitFor;
+
+/**
+ * @author homoroselaps
+ */
+
+public abstract class BehaviorTreeFactory implements Serializable {
+    private static final long serialVersionUID = 8396039806339873520L;
+
+    protected static Node<Entity> WaitForTargetReached_FailIfNot() {
+        // wait for targetReached, but abort if target not reachable
+        return new Sequence<>(
+            new Inverter<>(TriggerCondition(SteeringComponent.TargetNotReachedTrigger.class)),
+            new WaitFor(SteeringComponent.TargetReachedTrigger.class, true)
+        );
+    }
+
+    protected static Guard<Entity> TriggerGuard(Class<? extends Notification> type, Node<Entity> child) {
+        return new Guard<>(entity -> entity.getNotificationsIt(type).hasNext(), true, child);
+    }
+
+    protected static Action<Entity> StartAnimation(EMovableAction animation, short duration) {
+        return new Action<>(e -> {
+            e.aniC().startAnimation(animation, duration);
+        });
+    }
+
+    protected static void convertTo(Entity entity, EMovableType type) {
+        Entity blueprint = EntityFactory.CreateEntity(entity.gameC().getMovableGrid(), type, entity.movC().getPos(), entity.movC().getPlayer());
+        entity.convertTo(blueprint);
+    }
+
+    protected static <T> Node<T> Optional(Node<T> child) {
+        return new Selector<T>(child, new Succeeder<T>());
+    }
+
+    protected static class Wait extends Node<Entity> {
+        private static final long serialVersionUID = 8774557186392581042L;
+        boolean done = false;
+        int delay = -1;
+        public Wait(int milliseconds) {
+            super();
+            delay = milliseconds;
+        }
+
+        @Override
+        protected NodeStatus onTick(Tick<Entity> tick) {
+            if (done) return NodeStatus.Success;
+            tick.Target.setInvocationDelay(delay);
+            done = true;
+            return NodeStatus.Running();
+        }
+
+        @Override
+        protected void onOpen(Tick<Entity> tick) {
+            done = false;
+        }
+    }
+
+    protected static Condition<Entity> TriggerCondition(Class<? extends Notification> type) {
+        return new Condition<>(entity -> entity.getNotificationsIt(type).hasNext());
+    }
+
+    protected static Debug $(String message, Node<Entity> child) {
```

Comment:
The name of the method does not express its behaviour.

Reply:
Thank you for your annotation. You are completely right, this is indicates bad code style.
In this decision I prioritized readability of the whole behavior tree over comprehensibility of this single function. Some behavior tree frameworks ship with a graphical tool, which we don't have. Therefore, I did my best to avoid as much boilerplate code/code repetition as possible, to support behavior design in code.
The function $ is here defined as a shorthand for the creation of a ´Debug´ node, which annotates tree nodes with debug information/comments.
I choose $ as to imitate the jQuery function, which developers might already be familiar with.
I'd fine with any other valid symbol, but would rather not clatter the tree with to many 
```new Debug(...)``` statements.